src/main/java/net/pms/network/api/ApiHandler.java

```diff
@@ -0,0 +1,136 @@
+package net.pms.network.api;
+
+import java.io.File;
+import java.nio.file.Path;
+import java.util.List;
+import java.util.Map.Entry;
+import org.jboss.netty.channel.MessageEvent;
+import org.jboss.netty.handler.codec.http.HttpHeaders;
+import org.jboss.netty.handler.codec.http.HttpMethod;
+import org.jboss.netty.handler.codec.http.HttpRequest;
+import org.jboss.netty.handler.codec.http.HttpResponse;
+import org.jboss.netty.handler.codec.http.HttpResponseStatus;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import net.pms.PMS;
+import net.pms.configuration.RendererConfiguration;
+import net.pms.dlna.DLNAMediaDatabase;
+import net.pms.dlna.DLNAResource;
+import net.pms.dlna.RealFile;
+import net.pms.dlna.RootFolder;
+
+public class ApiHandler {
+
+	private static final Logger LOGGER = LoggerFactory.getLogger(ApiHandler.class);
+
+	private DLNAMediaDatabase database = PMS.get().getDatabase();
+
+	public ApiHandler() {
+	}
+
+	/**
+	 * Handle API calls
+	 *
+	 * @param method HTTP method
+	 * @param content body content
+	 * @param output Response to request
+	 * @param event The request
+	 * @param uri
+	 */
+	public void handleApiRequest(HttpMethod method, String content, HttpResponse output, String uri, MessageEvent event) {
+		output.headers().set(HttpHeaders.Names.CONTENT_LENGTH, "0");
+		output.setStatus(HttpResponseStatus.NO_CONTENT);
+
+		String serverApiKey = PMS.getConfiguration().getApiKey();
+		if (serverApiKey.length() < 12) {
+			LOGGER.warn("Weak server API key configured. UMS.conf api_key should have at least 12 digests.");
+			output.setStatus(HttpResponseStatus.SERVICE_UNAVAILABLE);
+			return;
+		}
+
+		String givenApiKey = extractApiKey(event);
+		if (serverApiKey.equals(givenApiKey)) {
```

Comment:
If this API handler is used for more important things in the future, there could be a benefit of making this API authentication more robust - thought it might also be overkill for UMS. Using the string comparison can be used for timing attacks especially since this is a cleartext comparison (String.equals(String) returns as soon as a single character doesn't match so it can be used to brute force passwords)

Could maybe do something like 

```
public static boolean arePasswordsEqual(String a, String b) {
    boolean equal = true;
    if ( a.length != b.length ) {
       equal = false;
    }

    for ( int i = 0; i < MAX_PASSWORD_LENGTH; i++ ) {
        if ( a.charAt(i%a.length()) != b.charAt(i%b.length()) ) {
            equal = false;
        }
    }
    return equal;
 }
```

Reply:
Might as well do it right the first time, instead of having to remember to "upgrade" such logic later should the need arise.

I have a couple of questions to your approach though:

- Why would you want to do the initial length comparison? That will allow a time attack to easily determine the password length.
- Why the strange constructs with `MAX_PASSWORD_LENGTH` and `i%x.length()`? Isn't it easier to just do `i < Math.min(a.length, b.length)` in the loop definition and then do the length comparison as the very last step before returning?

Reply:
I'm sure the brains here can come up with a much better implementation - I was more keen to point out the timing attack possibility with `String.equals(`. Your suggestions sound good

Reply:
Hi @js-kyle . Thanks for pointing out this issue. I've updated the code to prevent timing attacks by hashing the "given" and "server" api keys and always comparing the complete byte array.